kernel/appenv.c

```diff
@@ -0,0 +1,252 @@
+        }
+    }
+
+    result = JSON_OK;
+
+done:
+    return result;
+}
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstack-usage="
+int myst_create_appenv(myst_kernel_args_t* args)
+{
+    int ret = 0;
+    size_t size;
+    char* data = NULL;
+    char* path = APPENV_FILENAME;
+
+    /* open the APPENV_FILENAME */
+    ECHECK(myst_load_file(path, (void**)&data, &size));
+
+    /* read all the env vars and values */
+    json_parser_t parser;
+    json_result_t r;
+    json_callback_data_t callback_data;
+    static json_allocator_t allocator = {
+        malloc,
+        free,
+    };
+
+    memset(&callback_data, 0, sizeof(json_callback_data_t));
+    myst_args_init(&callback_data.env);
+
+    const json_parser_options_t options = {1};
+
+    if ((r = json_parser_init(
+             &parser,
+             data,
+             size,
+             _json_read_callback,
+             &callback_data,
+             &allocator,
+             &options)) != JSON_OK)
+    {
+        myst_eprintf("json_parser_init() failed: %d\n", r);
+        ERAISE(-EINVAL);
+    }
+
+    parser.trace = _trace;
+
+    if ((r = json_parser_parse(&parser)) != JSON_OK)
+    {
+        myst_eprintf("json_parser_parse() failed: %d\n", r);
+        ERAISE(-EINVAL);
+    }
+
+    if (parser.depth != 0)
+    {
+        myst_eprintf("unterminated objects\n");
+        ERAISE(-EINVAL);
+    }
+
+    if (callback_data.env.size == 0)
+    {
+        ERAISE(-EINVAL);
+    }
+
+    myst_args_t new_envp;
+    size_t new_envc = args->envc;
+
+    if (myst_args_init(&new_envp) != 0)
+        ERAISE(-ENOMEM);
+
+    /* store a copy of all the current env vars in new_env */
+    for (size_t i = 0; i < args->envc; i++)
+    {
+        char* nv = NULL;
+        if (asprintf(&nv, "%s", args->envp[i]) == -1)
+            ERAISE(-ENOMEM);
+
+        if (myst_args_append1(&new_envp, nv) != 0)
+            ERAISE(-EINVAL);
+    }
+
+    /* store the appenv variables in the env vector according to rules */
+    for (size_t i = 0; i < callback_data.env.size; i++)
+    {
+        char* nv = NULL;
+        if (asprintf(&nv, "%s", callback_data.env.data[i]) == -1)
+            ERAISE(-ENOMEM);
+
+        const char* value = strchr(nv, '=');
+        if (!value)
+        {
+            myst_eprintf("Environment variable does not have '='\n");
+            ERAISE(-EINVAL);
+        }
+
+        size_t name_len = value - nv;
+        value++;
+
+        /* if environment variable is not already defined, add it to envp */
+        int pos;
+        if ((pos = myst_args_find(&new_envp, nv, name_len)) <= 0)
+        {
+            if (myst_args_append1(&new_envp, nv) != 0)
+                ERAISE(-EINVAL);
+            new_envc++;
+        }
+        else
+        {
+            /* if the environment variable is already defined, skip or concat to
+             * its current value if it is PATH or LD_LIBRARY_PATH */
+            if (strncmp(new_envp.data[pos], "PATH=", 5) == 0)
```

Comment:
We can combine this if-statement and the next one. Also try not to use literal integer for literal string lengths otherwise we will easily screw up when later we modify the string but forget the length. The better way is:

```c
const char* pattern = "....";
if (strncmp(longer_string, pattern, strlen(pattern)) == 0)
   ...
```

Reply:
Thank you for the suggestion, added